go/core/action.go

```diff
@@ -28,6 +28,26 @@ import (
 	"github.com/invopop/jsonschema"
 )
 
+// Middleware is a functions that take in the request object, response object,
+// and the next middleware function for the action.
+//
+// Middleware functions can:
+//   - execute arbitrary code;
+//   - change the request and response;
+//   - terminate response by returning a response (or error);
+//   - call the next middleware function.
+//
+// If the current middleware function does not terminate the middlware chain it
+// must call next middlware in the chain to pass control to the next middleware
+// function, otherwise the request will be left hanging.
+type Middleware[I, O any] func(context.Context, I, func(ctx context.Context, input I) (O, error)) (O, error)
```

Comment:
Passing in the next function to call works, but I think more typical would be to return an error like `ErrMiddlewareDone`.  If the caller sees that error, it will return the output without continuing to call middleware or the model.

Reply:
There is only one caller of the middleware functions as far as I can see, in `Action.Run`.  I meant that that function could loop through the middleware slice and call each function.  If one of them returned `ErrMiddlewareDone`, it would break out of the loop.

Reply:
You might be thinking of a different pattern. I refactored the interface a bit, following an existing Go middleware pattern implementation https://cs.opensource.google/go/x/pkgsite/+/d8c29b80:internal/middleware/middleware.go

Reply:
I think the main point I was missing is that this middleware can transform the output, so it needs to run both before and after the other code.  Thanks.